cache/cache.go

```diff
@@ -13,3 +13,192 @@
 // limitations under the License.
 
 package cache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	rpctypes "go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+var ErrCompacted = rpctypes.ErrCompacted
+
+// WatchEventStream is the read-only channel returned to a single Watch() caller.
+type WatchEventStream <-chan *clientv3.Event
+
+// Cache buffers a single etcd Watch for a given key‐prefix and fan‑outs local watchers.
+// It is created internally by ShardedCache; callers typically use ShardedCache instead.
+type Cache struct {
+	prefix  string             // prefix is the key-prefix this shard is responsible for ("" = root).
+	cfg     Config             // immutable runtime configuration
+	client  *clientv3.Client   // etcd client used for upstream calls
+	history History            // history stores recent events in a fixed-size ring and supports replay.
+	demux   *demux             // demux fans incoming events out to active watchers and manages resync.
+	ready   chan struct{}      // closed once the upstream-watch goroutine is running
+	stop    context.CancelFunc // shuts down all cache goroutines
+	wg      sync.WaitGroup     // waits for goroutines to exit
+
+}
+
+// NewWithPrefix builds a cache shard that watches only the requested prefix.
+// For the root cache pass "".
+func NewWithPrefix(ctx context.Context, client *clientv3.Client, prefix string, opts ...Option) (*Cache, error) {
```

Comment:
Nit: Any reason to name it `NewWithPrefix` are you planning to have a `New` without prefix? If empty prefix works semantically correct then I don't see a reason to plan for such a function.